homeassistant/components/sensor/recorder.py

```diff
@@ -427,34 +428,41 @@ def _compile_statistics(  # noqa: C901
             entity_ids=entities_significant_history,
         )
         history_list = {**history_list, **_history_list}
-    # If there are no recent state changes, the sensor's state may already be pruned
-    # from the recorder. Get the state from the state machine instead.
-    for _state in sensor_states:
-        if _state.entity_id not in history_list:
-            history_list[_state.entity_id] = [_state]
 
-    to_process = []
-    to_query = []
+    entities_with_float_states: dict[str, list[tuple[float, State]]] = {}
     for _state in sensor_states:
         entity_id = _state.entity_id
-        if entity_id not in history_list:
+        # If there are no recent state changes, the sensor's state may already be pruned
+        # from the recorder. Get the state from the state machine instead.
+        if not (entity_history := history_list.get(entity_id, [_state])):
             continue
+        if not (float_states := _entity_history_to_float_and_state(entity_history)):
+            continue
+        entities_with_float_states[entity_id] = float_states
 
-        entity_history = history_list[entity_id]
-        statistics_unit, fstates = _normalize_states(
+    # Only lookup metadata for entities that have valid float states
+    # since it will result in cache misses for statistic_ids
+    # that are not in the metadata table and we are not working
+    # with them anyway.
+    old_metadatas = statistics.get_metadata_with_session(
```

Comment:
I had to refactor this a bit to valid the states are float states before fetching the metadata as we would otherwise look for metadata for statistics we were never going to generate